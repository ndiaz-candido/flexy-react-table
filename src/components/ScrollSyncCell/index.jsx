import React from 'react';
import PropTypes from 'prop-types';

const ScrollSyncCell = ({ column }) => (<div>Cell rendering content for { column }</div>);

ScrollSyncCell.propTypes = {
  /** The column name where the cell should be rendered */
  column: PropTypes.string.isRequired
};

export default ScrollSyncCell;
