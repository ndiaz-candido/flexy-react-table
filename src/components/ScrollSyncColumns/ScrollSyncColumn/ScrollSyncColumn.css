.scrollSyncColumn {
  padding: 10px;
}
