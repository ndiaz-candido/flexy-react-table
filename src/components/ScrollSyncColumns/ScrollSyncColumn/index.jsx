import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { scrollSyncColumn } from './index.module.css';

const ScrollSyncColumn = ({ width, children, className }) => {
  const styles = {
    width,
    flex: `0 0 ${width}`,
  };
  const classNames = classnames(className, scrollSyncColumn);

  return (
    <div className={classNames} style={styles}>
      {children}
    </div>
  );
};

ScrollSyncColumn.propTypes = {
  width: PropTypes.string,
  children: PropTypes.any.isRequired,

  /** The class name to override the default styles */
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
};

ScrollSyncColumn.defaultProps = {
  width: '150px',
  className: ''
};

export default ScrollSyncColumn;
