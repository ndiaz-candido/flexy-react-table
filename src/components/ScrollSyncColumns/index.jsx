import React from 'react';

const ScrollSyncColumns = () => <div />;

export default ScrollSyncColumns;
