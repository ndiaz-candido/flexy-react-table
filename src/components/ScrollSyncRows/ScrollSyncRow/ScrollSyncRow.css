.scrollSyncRow {
  display: flex;
  overflow-x: hidden;
  min-height: 50px;
}

.sticky {
  position: sticky;
  top: 0;
  z-index: 1000;
}
