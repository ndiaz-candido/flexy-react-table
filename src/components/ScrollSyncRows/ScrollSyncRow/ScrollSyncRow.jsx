import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  LeftStickySection,
  RightStickySection,
  ScrollableSection,
} from './sections';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};
const cellDefaultStyles = width => ({
  width,
  flex: `0 0 ${width}`,
});

class ScrollSyncRow extends PureComponent {
  static DEFAULT_COLUMN_WIDTH = 50;
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  rowId = -1;

  constructor(props) {
    super(props);

    this.rowId = Number(props.rowId);
  }

  state = {
    leftStickySection: [],
    rightStickySection: [],
    scrollableSection: [],
  };

  static getDerivedStateFromProps({
    columns,
    children = [],
    columnClassName,
    isHeader,
  }) {
    const leftStickySection = [];
    const rightStickySection = [];
    const scrollableSection = [];

    columns.forEach(column => {
      const {
        dataKey = '',
        width = ScrollSyncRow.DEFAULT_COLUMN_WIDTH,
        stickyAlign,
      } = column.props;
      const cellContent = isHeader
        ? column.props.children
        : React.Children.toArray(children).find(
            cell => cell.props.column === dataKey
          );
      const styles = cellDefaultStyles(width);
      const cell = (
        <div className={columnClassName} style={styles}>
          {cellContent}
        </div>
      );

      switch (stickyAlign) {
        case 'left':
          leftStickySection.push(cell);
          break;
        case 'right':
          rightStickySection.push(cell);
          break;
        default:
          scrollableSection.push(cell);
          break;
      }
    });

    return {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    };
  }

  render() {
    const { isSticky, scrollLeft, onScroll, className } = this.props;
    const {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    } = this.state;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftStickySection cells={leftStickySection} />
        <ScrollableSection
          rowId={this.rowId}
          cells={scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightStickySection cells={rightStickySection} />
      </div>
    );
  }
}

ScrollSyncRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
};

ScrollSyncRow.defaultProps = {
  className: '',
};

export default ScrollSyncRow;
