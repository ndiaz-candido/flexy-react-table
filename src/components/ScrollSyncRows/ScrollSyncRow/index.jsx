import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import ScrollSyncTableCell from '../../ScrollSyncCell';
import LeftStickySection from './Sections/LeftStickySection';
import RightStickySection from './Sections/RightStickySection';
import ScrollableSection from './Sections/ScrollableSection';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};

class ScrollSyncRow extends PureComponent {
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  rowColumns = [];
  leftStickySection = [];
  rightStickySection = [];
  scrollableSection = [];
  rowId = -1;

  constructor(props) {
    super(props);

    this.rowId = Number(props.rowId);
  }

  componentWillMount() {
    if (this.rowId === ScrollSyncRow.HEADER_ROW_ID) {
      this.rowColumns = this.injectPropsToHeaderColumns();
      this.distribuiteColumnsPerSection();
    } else if (this.areChildrenSyncCells()) {
      this.rowColumns = this.setChildPerColumn();
      this.distribuiteColumnsPerSection();
    }
  }

  areChildrenSyncCells = () => {
    const { children } = this.props;
    const childrenArray = React.Children.toArray(children);

    return childrenArray.some(child => child.type === ScrollSyncTableCell);
  };

  injectPropsToRegularColumn = column => {
    const { children, columnClassName } = this.props;
    const columnName = column.props.name;
    const scrollSyncCells = React.Children.toArray(children);
    const scrollSyncCellMatchingColumn = scrollSyncCells.find(
      cell => cell.props.column === columnName
    );
    const scrollSyncCellMatchingColumnChildren = scrollSyncCellMatchingColumn
      ? scrollSyncCellMatchingColumn.props.children
      : '';

    return React.cloneElement(column, {
      children: scrollSyncCellMatchingColumnChildren,
      className: columnClassName,
    });
  };

  injectPropsToHeaderColumns = () => {
    const { columns, columnClassName } = this.props;

    return columns.map(column =>
      React.cloneElement(column, {
        className: columnClassName,
      })
    );
  };

  setChildPerColumn = () => {
    const { children, columns } = this.props;
    let rowColumns = [];

    if (children && React.Children.count(children) > 0) {
      rowColumns = columns.map(this.injectPropsToRegularColumn);
    }

    return rowColumns;
  };

  distribuiteColumnsPerSection = () => {
    this.rowColumns.forEach(column => {
      switch (column.props.stickyAlign) {
        case 'left':
          this.leftStickySection.push(column);
          break;
        case 'right':
          this.rightStickySection.push(column);
          break;
        default:
          this.scrollableSection.push(column);
          break;
      }
    });
  };

  render() {
    const { isSticky, scrollLeft, onScroll, className } = this.props;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftStickySection columns={this.leftStickySection} />
        <ScrollableSection
          rowId={this.rowId}
          columns={this.scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightStickySection columns={this.rightStickySection} />
      </div>
    );
  }
}

ScrollSyncRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
};

ScrollSyncRow.defaultProps = {
  className: '',
};

export default ScrollSyncRow;
