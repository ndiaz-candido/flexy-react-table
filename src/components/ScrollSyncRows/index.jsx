import React from 'react';

const ScrollSyncRows = () => <div />;

export default ScrollSyncRows;
