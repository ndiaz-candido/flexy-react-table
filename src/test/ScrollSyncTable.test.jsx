import { shallow } from 'enzyme';
import ScrollSyncTable from '../components/ScrollSyncTable';
import ScrollSyncColumns from '../components/ScrollSyncColumns';
import ScrollSyncColumn from '../components/ScrollSyncColumns/ScrollSyncColumn';
import ScrollSyncRows from '../components/ScrollSyncRows';
import ScrollSyncRow from '../components/ScrollSyncRows/ScrollSyncRow';
import ScrollSyncCell from '../components/ScrollSyncCell';

describe('ScrollSyncTable', () => {
  test('should render with default props', () => {
    const ScrollSyncTableComponent = shallow(<ScrollSyncTable />);

    expect(ScrollSyncTableComponent).toMatchSnapshot();
  });

  test('should render with the given class names', () => {
    const ScrollSyncTableComponent = shallow(
      <ScrollSyncTable
        tableClassName="table-class-name"
        rowClassName="row-class-name"
        columnClassName="column-class-name"
        headerClassName="header-class-name"
        headerColumnClassName="heade-column-class-name"
      />
    );

    expect(ScrollSyncTableComponent).toMatchSnapshot();
  });

  test('should pass the prop stickHeader to the table body', () => {
    const ScrollSyncTableComponent = shallow(<ScrollSyncTable stickHeader />);

    expect(ScrollSyncTableComponent).toMatchSnapshot();
  });

  test('should render the given columns and rows', () => {
    const ScrollSyncTableComponent = shallow(
      <ScrollSyncTable>
        <ScrollSyncColumns>
          <ScrollSyncColumn name="a">A</ScrollSyncColumn>
        </ScrollSyncColumns>
        <ScrollSyncRows>
          <ScrollSyncRow>
            <ScrollSyncCell column="a">A / A</ScrollSyncCell>
          </ScrollSyncRow>
        </ScrollSyncRows>
      </ScrollSyncTable>
    );

    expect(ScrollSyncTableComponent).toMatchSnapshot();
  });

  test('should handle rendering when there are not columns neither rows', () => {
    const ScrollSyncTableComponent = shallow(
      <ScrollSyncTable>
        <ScrollSyncColumns />
        <ScrollSyncRows />
      </ScrollSyncTable>
    );

    expect(ScrollSyncTableComponent).toMatchSnapshot();
  });
});
