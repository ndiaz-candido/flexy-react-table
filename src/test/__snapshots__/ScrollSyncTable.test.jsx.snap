// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollSyncTable should handle rendering when there are not columns neither rows 1`] = `
<div
  className=""
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <ScrollSyncBody
    columnClassName=""
    columns={Array []}
    headerClassName=""
    headerColumnClassName=""
    rowClassName=""
    rows={Array []}
    stickHeader={false}
  />
</div>
`;

exports[`ScrollSyncTable should pass the prop stickHeader to the table body 1`] = `
<div
  className=""
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <ScrollSyncBody
    columnClassName=""
    columns={Array []}
    headerClassName=""
    headerColumnClassName=""
    rowClassName=""
    rows={Array []}
    stickHeader={true}
  />
</div>
`;

exports[`ScrollSyncTable should render the given columns and rows 1`] = `
<div
  className=""
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <ScrollSyncBody
    columnClassName=""
    columns={
      Array [
        <ScrollSyncColumn
          className=""
          name="a"
          width="150px"
        >
          A
        </ScrollSyncColumn>,
      ]
    }
    headerClassName=""
    headerColumnClassName=""
    rowClassName=""
    rows={
      Array [
        <ScrollSyncRow
          className=""
        >
          <ScrollSyncCell
            column="a"
          >
            A / A
          </ScrollSyncCell>
        </ScrollSyncRow>,
      ]
    }
    stickHeader={false}
  />
</div>
`;

exports[`ScrollSyncTable should render with default props 1`] = `
<div
  className=""
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <ScrollSyncBody
    columnClassName=""
    columns={Array []}
    headerClassName=""
    headerColumnClassName=""
    rowClassName=""
    rows={Array []}
    stickHeader={false}
  />
</div>
`;

exports[`ScrollSyncTable should render with the given class names 1`] = `
<div
  className="table-class-name"
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <ScrollSyncBody
    columnClassName="column-class-name"
    columns={Array []}
    headerClassName="header-class-name"
    headerColumnClassName="header-column-class-name"
    rowClassName="row-class-name"
    rows={Array []}
    stickHeader={false}
  />
</div>
`;
